### PR TITLE
Add linkstats to module display

### PR DESCRIPTION
**Pull request by JyeSmith**
This provides an easy way to check Linkstats from a module with a display.  PPM or MAVLink users without a handset should find this beneficial.

From the Home page DOWN and UP press will now cycle between Linkstats and Home display.  Long Enter/Right press can be used the enter the menu system.

Downlink RSSI2 can now be seen on Gemini Tx modules with a display.  Up and Downlink SNR2 will require and OTA packet change.  This will have to wait for ELRS V4.

image

### CODE REVIEW

File: src/lib/SCREEN/TFT/tftdisplay.cpp [hunk 1]
```diff
@@ -11,10 +11,13 @@
 #include "options.h"
 #include "logging.h"
 #include "common.h"
+#include "CRSF.h"
 
 #include "WiFi.h"
 extern WiFiMode_t wifiMode;
 
+extern CRSF crsf;
+
```
Comment (pkendall64) [inline src/lib/SCREEN/TFT/tftdisplay.cpp:20]: Not needed. The variable is static.

File: src/lib/SCREEN/TFT/tftdisplay.cpp [hunk 2]
```diff
@@ -382,4 +395,63 @@ void TFTDisplay::displaySending()
...
+    gfx->setCursor(LINKSTATS_COL_FIRST, LINKSTATS_ROW_THIRD);
+    gfx->print("RSSI");
+    gfx->setCursor(LINKSTATS_COL_FIRST, LINKSTATS_ROW_FOURTH);
+    gfx->print("SNR");
+    gfx->setCursor(LINKSTATS_COL_FIRST, LINKSTATS_ROW_FIFTH);
+    gfx->print("Ant");
+
+    // Uplink Linkstats
+    gfx->setCursor(LINKSTATS_COL_SECOND, LINKSTATS_ROW_SECOND);
+    snprintf(buffer, sizeof(buffer), "%03d", crsf.LinkStatistics.uplink_Link_quality);
```
Comment (pkendall64) [inline src/lib/SCREEN/TFT/tftdisplay.cpp:420]: Access these all with the the static accessor i.e. `CRSF::LinkStatistics.uplink_Link_quality` etc

File: src/lib/SCREEN/menu.cpp [hunk 3]
```diff
@@ -631,17 +635,32 @@ fsm_state_entry_t const main_menu_fsm[] = {
...
     {EVENT_TIMEOUT, GOTO(STATE_IDLE)},
     {EVENT_LONG_ENTER, PUSH(main_menu_fsm)},
-    {EVENT_LONG_RIGHT, PUSH(main_menu_fsm)}
+    {EVENT_LONG_RIGHT, PUSH(main_menu_fsm)},
+    {EVENT_UP, PUSH(linkstats_menu_fsm)},
+    {EVENT_DOWN, PUSH(linkstats_menu_fsm)}
 };
 fsm_state_entry_t const entry_fsm[] = {
-    {STATE_SPLASH, nullptr, displaySplashScreen, 5000, splash_events, ARRAY_SIZE(splash_events)},
+    {STATE_SPLASH, nullptr, displaySplashScreen, 1000, splash_events, ARRAY_SIZE(splash_events)},
```
Comment (pkendall64) [inline src/lib/SCREEN/menu.cpp:663]: Did you really mean to shorten this to 1 second? Or was it just for debugging?

Comment (JyeSmith) [reply src/lib/SCREEN/menu.cpp:663]: Debugging... but it would be nice to make it a little shorter. 3000?

Comment (pkendall64) [reply src/lib/SCREEN/menu.cpp:663]: I'm cool with that.

File: src/lib/CrsfProtocol/crsf_protocol.h [hunk 4]
```diff
@@ -364,7 +364,8 @@ typedef struct crsfPayloadLinkstatistics_s
     uint8_t active_antenna;
     uint8_t rf_Mode;
     uint8_t uplink_TX_Power;
-    uint8_t downlink_RSSI;
+    uint8_t downlink_RSSI_1;
+    uint8_t downlink_RSSI_2;
```
Comment (CapnBry) [inline src/lib/CrsfProtocol/crsf_protocol.h:368]: You can't just add a field to this structure I'm afraid, as this structure defines the CRSF on-the-wire packet. With this PR, the LinkStatistics going out as CRSF are broken. You'll have to store the data somewhere else.

File: src/lib/SCREEN/TFT/tftdisplay.cpp [hunk 5]
```diff
@@ -119,6 +120,16 @@ constexpr uint16_t elrs_banner_bgColor[] = {
...
+//LINKSTATS PAGE Definition
+#define LINKSTATS_COL_FIRST   0
+#define LINKSTATS_COL_SECOND  30
+#define LINKSTATS_COL_THIRD   100
+#define LINKSTATS_ROW_FIRST   10
+#define LINKSTATS_ROW_SECOND  25
+#define LINKSTATS_ROW_THIRD   40
+#define LINKSTATS_ROW_FOURTH  55
+#define LINKSTATS_ROW_FIFTH   70
+
```
Comment (CapnBry) [inline src/lib/SCREEN/TFT/tftdisplay.cpp:132]: These can just be constants in the `displayLinkstats()` function as that is the only place they are used and it keeps the code self-contained.

File: src/lib/SCREEN/OLED/oleddisplay.cpp [hunk 6]
```diff
@@ -396,6 +397,48 @@ void OLEDDisplay::displaySending()
     u8g2->sendBuffer();
 }
 
+void OLEDDisplay::displayLinkstats()
+{
+    u8g2->clearBuffer();
+    u8g2->setFont(u8g2_font_profont10_mr);
+
+    u8g2->drawStr(0, 20, "LQ");
```
Comment (CapnBry) [inline src/lib/SCREEN/OLED/oleddisplay.cpp:405]: TFT gets constants, OLED gets numbers?

File: src/lib/SCREEN/OLED/oleddisplay.cpp [hunk 7]
```diff
@@ -396,6 +397,48 @@ void OLEDDisplay::displaySending()
...
+    u8g2->drawStr(0, 30, "RSSI");
+    u8g2->drawStr(0, 40, "SNR");
+    u8g2->drawStr(0, 50, "Ant");
+
+    u8g2->drawStr(32, 10, "Uplink");
+    u8g2->setCursor(32, 20);
+    u8g2->print(CRSF::LinkStatistics.uplink_Link_quality);
+    u8g2->setCursor(32, 30);
+    u8g2->print((int8_t)CRSF::LinkStatistics.uplink_RSSI_1);
+    u8g2->print("/");
```
Comment (CapnBry) [inline src/lib/SCREEN/OLED/oleddisplay.cpp:415]: Maybe wrap RSSI_2 in something that only prints if it is non-zero?

I've also just realized if we switch from a diversity RX to a non-diversity RX, our linkstats will keep displaying RSSI_2 from the previous session. I am not sure how the heck we'd fix that though without zeroing out the values on disconnect which could be throwing away important information at range.

File: src/lib/SCREEN/TFT/tftdisplay.cpp [hunk 8]
```diff
@@ -382,4 +393,63 @@ void TFTDisplay::displaySending()
...
+    gfx->print("RSSI");
+    gfx->setCursor(LINKSTATS_COL_FIRST, LINKSTATS_ROW_FOURTH);
+    gfx->print("SNR");
+    gfx->setCursor(LINKSTATS_COL_FIRST, LINKSTATS_ROW_FIFTH);
+    gfx->print("Ant");
+
+    // Uplink Linkstats
+    gfx->setCursor(LINKSTATS_COL_SECOND, LINKSTATS_ROW_SECOND);
+    snprintf(buffer, sizeof(buffer), "%03d", CRSF::LinkStatistics.uplink_Link_quality);
+    gfx->print(String(buffer));
```
Comment (CapnBry) [inline src/lib/SCREEN/TFT/tftdisplay.cpp:419]: These String constructors are not needed, there is an `gfx->print(const char [])`  to use buffer, or even easier is to use `gfx->printf(...)` and not do any snprintfs

**Review by CapnBry — CHANGES_REQUESTED** [review]
Looks good and works but changes are needed to not break CRSF protocol.

File: src/lib/CrsfProtocol/crsf_protocol.h (hunk 4 above)
Comment (CapnBry) [reply src/lib/CrsfProtocol/crsf_protocol.h:368]: Not thrilled about doing this considering how things might later expect the struct to be a certain size, but the way our code is written now, this works (subclassing the wire struct to add a field). Also removes volatile because that is not needed and shaves off 84 bytes of instructions. Can push if you like.
```diff
diff --git a/src/lib/CrsfProtocol/crsf_protocol.h b/src/lib/CrsfProtocol/crsf_protocol.h
index 5a4c080e..077b41fd 100644
--- a/src/lib/CrsfProtocol/crsf_protocol.h
+++ b/src/lib/CrsfProtocol/crsf_protocol.h
@@ -364,12 +364,15 @@ typedef struct crsfPayloadLinkstatistics_s
     uint8_t active_antenna;
     uint8_t rf_Mode;
     uint8_t uplink_TX_Power;
-    uint8_t downlink_RSSI;
+    uint8_t downlink_RSSI_1;
     uint8_t downlink_Link_quality;
     int8_t downlink_SNR;
-} crsfLinkStatistics_t;
+} PACKED crsfLinkStatistics_t;

-typedef struct crsfPayloadLinkstatistics_s crsfLinkStatistics_t;
+typedef struct elrsLinkStatistics_s : crsfLinkStatistics_t
+{
+    uint8_t downlink_RSSI_2;
+} PACKED elrsLinkStatistics_t;

 // typedef struct crsfOpenTXsyncFrame_s
 // {
diff --git a/src/lib/Handset/CRSF.cpp b/src/lib/Handset/CRSF.cpp
index 8cdaac03..ea1eab58 100644
--- a/src/lib/Handset/CRSF.cpp
+++ b/src/lib/Handset/CRSF.cpp
@@ -3,7 +3,7 @@
 #include "common.h"
 #include "FIFO.h"

-volatile crsfPayloadLinkstatistics_s CRSF::LinkStatistics;
+elrsLinkStatistics_t CRSF::LinkStatistics;
 GENERIC_CRC8 crsf_crc(CRSF_CRC_POLY);

 uint8_t CRSF::MspData[ELRS_MSP_BUFFER] = {0};
diff --git a/src/lib/Handset/CRSF.h b/src/lib/Handset/CRSF.h
index 01eb208b..4bf3a2f3 100644
--- a/src/lib/Handset/CRSF.h
+++ b/src/lib/Handset/CRSF.h
@@ -9,7 +9,7 @@
 class CRSF
 {
 public:
-    static volatile crsfPayloadLinkstatistics_s LinkStatistics; // Link Statistics Stored as Struct
+    static elrsLinkStatistics_t LinkStatistics; // Link Statistics Stored as Struct

     static void GetMspMessage(uint8_t **data, uint8_t *len);
     static void UnlockMspMessage();
diff --git a/src/src/tx_main.cpp b/src/src/tx_main.cpp
index 84b616a3..821ff122 100644
--- a/src/src/tx_main.cpp
+++ b/src/src/tx_main.cpp
@@ -212,7 +212,7 @@ bool ICACHE_RAM_ATTR ProcessTLMpacket(SX12xxDriverCommon::rx_status const status

   Radio.GetLastPacketStats();
   CRSF::LinkStatistics.downlink_SNR = SNR_DESCALE(Radio.LastPacketSNRRaw);
-  CRSF::LinkStatistics.downlink_RSSI = Radio.LastPacketRSSI;
+  CRSF::LinkStatistics.downlink_RSSI_1 = Radio.LastPacketRSSI;

   // Full res mode
   if (OtaIsFullRes)
```

Comment (JyeSmith) [reply src/lib/CrsfProtocol/crsf_protocol.h:368]: >Can push these changes (and all the other ones I suggested) if you agree.

I just got sometime to sit down and go through all of this properly... then read this!

YES PLEASE, PUSH ALL THE THINGS!

Comment (JyeSmith) [reply src/lib/CrsfProtocol/crsf_protocol.h:368]: > You can't just add a field to this structure I'm afraid, as this structure defines the CRSF on-the-wire packet. With this PR, the LinkStatistics going out as CRSF are broken. You'll have to store the data somewhere else.

Yep, I totally fucked this.  I was to busy looking at OTA packets :sweat_smile:

Comment (CapnBry) [reply src/lib/CrsfProtocol/crsf_protocol.h]: > Yep, I totally fucked this. I was to busy looking at OTA packets 😅

That's what I thought. Jye checking out the back of his module admiring how cool it looks, while on the other side the linkstats are displaying all weird on the handset 😄

**Comment by CapnBry** [comment]
Pushed! I almost got a little carried away getting rid of the disconnected "frame length" constants that were kinda sorta used to track the size of the linkstats / channels frames.

Tested on
* TX16S internal (for incoming linkstats reporting over CRSF only) + iFlight 2.4 RX and RadioMaster RP4 RX
* Axisflying Thor TX (TFT) + iFlight 2.4 RX and RadioMaster RP4 RX
* RadioMaster Ranger Bling (OLED) + iFlight 2.4 RX and RadioMaster RP4 RX

In the TFT code, the use of the 0 in the formats you've used means all the values are 0-padded but on the OLED they are not. Are the extra 0s intentional? The antenna looks weird with a 0 and LQ with a leading 0? I'm fine with it if that's how you want it to look but I can make it match the OLED if this was unintentional.
image

**Comment by JyeSmith** [comment]
@CapnBry not intentional and if you have something ready to push, please do then we can get this merged. 

If not, no probs and I'll get it done today.

**Comment by CapnBry** [comment]
> @CapnBry not intentional and if you have something ready to push, please do then we can get this merged.

Okie doke, I've updated the TFT code to match the OLED code. Since they're very the same now, I also reordered the TFT code to match the order of the OLED code so they're visually similar too.
image

**Review by CapnBry — APPROVED** [review]
I'm good if you're good with my changes! If not, let me know if I can help meet your approval.
